### PR TITLE
Allow downloading sample data as a spreadsheet

### PR DESCRIPTION
**Pull request by apmasell**

### CODE REVIEW

File: miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/util/MisoWebUtils.java [hunk 1]
```diff
@@ -109,4 +121,23 @@ private static boolean checkFile(File path) throws IOException {
     }
     return checks;
   }
+
+  private static Pattern COMMA = Pattern.compile(",");
```
Comment (sqwci) [inline miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/util/MisoWebUtils.java]: MINOR Rename this field "COMMA" to match the regular expression '^[a-z][a-zA-Z0-9]*$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3008)

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/spreadsheet/SpreadSheetCollector.java [hunk 2]
```diff
@@ -15,34 +16,34 @@
 /**
  * Collect a stream of string rows into a spreadsheet.
  */
-public abstract class SpreadSheetCollector<Workbook, Sheet, Row>
-    implements Collector<String[], SpreadSheetCollector<Workbook, Sheet, Row>.State, byte[]> {
-  public class State {
+public abstract class SpreadSheetCollector<T, Workbook, Sheet, Row>
```
Comment (sqwci) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/spreadsheet/SpreadSheetCollector.java]: MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00119)
MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00119)
MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00119)

File: core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/spreadsheet/SpreadSheetCollector.java [hunk 3]
```diff
@@ -15,34 +16,34 @@
...
+    implements Collector<T, SpreadSheetCollector.State<Sheet, Workbook>, byte[]> {
+  public static class State<Sheet, Workbook> {
```
Comment (sqwci) [inline core/src/main/java/uk/ac/bbsrc/tgac/miso/core/data/spreadsheet/SpreadSheetCollector.java]: MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00119)
MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS00119)

**Comment by sqwci** [comment]
SonarQube analysis reported 11 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 8 minor

Watch the comments in this conversation to review them.

#### 5 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [MisoWebUtils.java#L106](https://github.com/TGAC/miso-lims/blob/0e61b9561c7bb0cb0287cbd76d8387b19b5477fb/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/util/MisoWebUtils.java#L106): Use the built-in formatting to construct this argument. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2629)
1. ![MAJOR][MAJOR] [hot.js#L51](https://github.com/TGAC/miso-lims/blob/0e61b9561c7bb0cb0287cbd76d8387b19b5477fb/miso-web/src/main/webapp/scripts/hot.js#L51): Refactor this code so that this expression does not always evaluate to true. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2589)
1. ![MAJOR][MAJOR] [hot.js#L228](https://github.com/TGAC/miso-lims/blob/0e61b9561c7bb0cb0287cbd76d8387b19b5477fb/miso-web/src/main/webapp/scripts/hot.js#L228): Refactor this code so that this expression does not always evaluate to true. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2589)
1. ![MINOR][MINOR] [hot_sample.js#L597](https://github.com/TGAC/miso-lims/blob/0e61b9561c7bb0cb0287cbd76d8387b19b5477fb/miso-web/src/main/webapp/scripts/hot_sample.js#L597): Remove this usage of alert(...). [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS1442)
1. ![MINOR][MINOR] [hot_sample.js#L602](https://github.com/TGAC/miso-lims/blob/0e61b9561c7bb0cb0287cbd76d8387b19b5477fb/miso-web/src/main/webapp/scripts/hot_sample.js#L602): Remove this usage of alert(...). [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS1442)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by sqwci** [comment]
SonarQube analysis reported 10 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 7 minor

Watch the comments in this conversation to review them.

#### 5 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [MisoWebUtils.java#L106](https://github.com/TGAC/miso-lims/blob/3b70e1a3f7bb77bab68c38d5514c2017c2cf7f84/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/util/MisoWebUtils.java#L106): Use the built-in formatting to construct this argument. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2629)
1. ![MAJOR][MAJOR] [hot.js#L51](https://github.com/TGAC/miso-lims/blob/3b70e1a3f7bb77bab68c38d5514c2017c2cf7f84/miso-web/src/main/webapp/scripts/hot.js#L51): Refactor this code so that this expression does not always evaluate to true. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2589)
1. ![MAJOR][MAJOR] [hot.js#L228](https://github.com/TGAC/miso-lims/blob/3b70e1a3f7bb77bab68c38d5514c2017c2cf7f84/miso-web/src/main/webapp/scripts/hot.js#L228): Refactor this code so that this expression does not always evaluate to true. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2589)
1. ![MINOR][MINOR] [hot_sample.js#L597](https://github.com/TGAC/miso-lims/blob/3b70e1a3f7bb77bab68c38d5514c2017c2cf7f84/miso-web/src/main/webapp/scripts/hot_sample.js#L597): Remove this usage of alert(...). [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS1442)
1. ![MINOR][MINOR] [hot_sample.js#L602](https://github.com/TGAC/miso-lims/blob/3b70e1a3f7bb77bab68c38d5514c2017c2cf7f84/miso-web/src/main/webapp/scripts/hot_sample.js#L602): Remove this usage of alert(...). [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS1442)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by callunity** [comment]
I tried downloading a tracking spreadsheet by selecting items in a box, but the Location column came out entirely blank. On any of the items' pages it shows the "Location" field as empty, but does display a value for "Box Location". I would have thought it would display the Box Location field; is that mistaken?

**Comment by apmasell** [comment]
Fixed.

**Comment by sqwci** [comment]
SonarQube analysis reported 10 issues
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 7 minor

Watch the comments in this conversation to review them.

#### 5 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [MisoWebUtils.java#L106](https://github.com/TGAC/miso-lims/blob/889e8482612531be0473f75c87f5061fdca52ac1/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/util/MisoWebUtils.java#L106): Use the built-in formatting to construct this argument. [rule](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2629)
1. ![MAJOR][MAJOR] [hot.js#L51](https://github.com/TGAC/miso-lims/blob/889e8482612531be0473f75c87f5061fdca52ac1/miso-web/src/main/webapp/scripts/hot.js#L51): Refactor this code so that this expression does not always evaluate to true. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2589)
1. ![MAJOR][MAJOR] [hot.js#L228](https://github.com/TGAC/miso-lims/blob/889e8482612531be0473f75c87f5061fdca52ac1/miso-web/src/main/webapp/scripts/hot.js#L228): Refactor this code so that this expression does not always evaluate to true. [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS2589)
1. ![MINOR][MINOR] [hot_sample.js#L597](https://github.com/TGAC/miso-lims/blob/889e8482612531be0473f75c87f5061fdca52ac1/miso-web/src/main/webapp/scripts/hot_sample.js#L597): Remove this usage of alert(...). [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS1442)
1. ![MINOR][MINOR] [hot_sample.js#L602](https://github.com/TGAC/miso-lims/blob/889e8482612531be0473f75c87f5061fdca52ac1/miso-web/src/main/webapp/scripts/hot_sample.js#L602): Remove this usage of alert(...). [rule](https://sonarcloud.io/coding_rules#rule_key=javascript%3AS1442)

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**